Build fails at test_setup_helpers.py with AttributeError

I can't build the AUR package python-extension-helpers on a fresh Archlinux install. I'm getting the following error output:

```
=========================================== FAILURES ===========================================
_____________________________________ test_compiler_module _____________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f0e497ac5e0>
c_extension_test_package = local('/tmp/pytest-of-tan/pytest-2/test_compiler_module0/test_pkg')

    def test_compiler_module(capsys, c_extension_test_package):
        """
        Test ensuring that the compiler module is built and installed for packages
        that have extension modules.
        """
    
        test_pkg = c_extension_test_package
        install_temp = test_pkg.mkdir('install_temp')
    
        with test_pkg.as_cwd():
            # This is one of the simplest ways to install just a package into a
            # test directory
>           run_setup('setup.py',
                      ['install',
                       '--single-version-externally-managed',
                       '--install-lib={0}'.format(install_temp),
                       '--record={0}'.format(install_temp.join('record.txt'))])

build/lib/extension_helpers/tests/test_setup_helpers.py:159: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

setup_script = 'setup.py'
args = ['install', '--single-version-externally-managed', '--install-lib=/tmp/pytest-of-tan/pytest-2/test_compiler_module0/test_pkg/install_temp', '--record=/tmp/pytest-of-tan/pytest-2/test_compiler_module0/test_pkg/install_temp/record.txt']

    def run_setup(setup_script, args):
    
        # This used to call setuptools.sandbox's run_setup, but due to issues with
        # this and Cython (which caused segmentation faults), we now use subprocess.
    
        setup_script = os.path.abspath(setup_script)
    
        path = os.path.dirname(setup_script)
        setup_script = os.path.basename(setup_script)
    
        if HAS_COVERAGE:
    
            # In this case, we run the command using the coverage command and we
            # then collect the coverage data into a SUBPROCESS_COVERAGE list which
            # is set up at the start of the testing process and is then combined
            # into a single .coverage file at the end of the testing process.
    
            p = sp.Popen(['coverage', 'run', setup_script] + list(args), cwd=path,
                         stdout=sp.PIPE, stderr=sp.PIPE)
            stdout, stderr = p.communicate()
    
            cdata = CoverageData()
>           cdata.read_file(os.path.join(path, '.coverage'))
E           AttributeError: 'CoverageData' object has no attribute 'read_file'

build/lib/extension_helpers/tests/__init__.py:68: AttributeError
======================================= warnings summary =======================================
../../../../../../../usr/lib/python3.9/site-packages/packaging/version.py:127: 1286 warnings
  /usr/lib/python3.9/site-packages/packaging/version.py:127: DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=================================== short test summary info ====================================
FAILED build/lib/extension_helpers/tests/test_setup_helpers.py::test_compiler_module - Attrib...
========================== 1 failed, 3 passed, 1286 warnings in 2.90s ==========================
Traceback (most recent call last):
  File "/usr/bin/pytest", line 33, in <module>
    sys.exit(load_entry_point('pytest==6.2.1', 'console_scripts', 'pytest')())
  File "/usr/lib/python3.9/site-packages/_pytest/config/__init__.py", line 185, in console_main
    code = main()
  File "/usr/lib/python3.9/site-packages/_pytest/config/__init__.py", line 162, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "/usr/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/usr/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/usr/lib/python3.9/site-packages/_pytest/main.py", line 316, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/usr/lib/python3.9/site-packages/_pytest/main.py", line 311, in wrap_session
    config._ensure_unconfigure()
  File "/usr/lib/python3.9/site-packages/_pytest/config/__init__.py", line 987, in _ensure_unconfigure
    self.hook.pytest_unconfigure(config=self)
  File "/usr/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/usr/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 208, in _multicall
    return outcome.get_result()
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/usr/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/tan/.cache/yay/python-extension-helpers/src/extension-helpers-0.1/build/lib/extension_helpers/conftest.py", line 57, in pytest_unconfigure
    combined_cdata.write_file('.coverage.subprocess')
AttributeError: 'CoverageData' object has no attribute 'write_file'
==> ERROR: A failure occurred in check().
    Aborting...
error making: python-extension-helpers

```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,4 +1,10 @@
+0.2 (unreleased)
+----------------
+
+* Added support for coverage>=5 for the extension-helpers test suite. [#24]
+
+
 0.1 (2019-12-18)
 ----------------
 
-Initial release of extension-helpers, which was forked from astropy-helpers 4.0.
+* Initial release of extension-helpers, which was forked from astropy-helpers 4.0.

--- a/extension_helpers/_openmp_helpers.py
+++ b/extension_helpers/_openmp_helpers.py
@@ -32,6 +32,7 @@ try:
     _EXTENSION_HELPERS_DISABLE_OPENMP_SETUP_
 except NameError:
     import builtins
+
     # It hasn't, so do so.
     builtins._EXTENSION_HELPERS_DISABLE_OPENMP_SETUP_ = False
 

--- a/extension_helpers/conftest.py
+++ b/extension_helpers/conftest.py
@@ -8,14 +8,16 @@ import os
 import glob
 
 try:
-    from coverage import CoverageData
+    from coverage import CoverageData, __version__ as coverage_version
 except ImportError:
     HAS_COVERAGE = False
+    CoverageData = None
 else:
-    HAS_COVERAGE = True
+    # Set to the major version number
+    HAS_COVERAGE = int(coverage_version.split('.')[0])
 
-if HAS_COVERAGE:
-    SUBPROCESS_COVERAGE = []
+
+SUBPROCESS_COVERAGE = []
 
 
 def pytest_configure(config):
@@ -26,9 +28,6 @@ def pytest_configure(config):
 def pytest_unconfigure(config):
 
     if HAS_COVERAGE:
-
-        # We create an empty coverage data object
-        combined_cdata = CoverageData()
 
         # Add all files from extension_helpers to make sure we compute the total
         # coverage, not just the coverage of the files that have non-zero
@@ -52,6 +51,15 @@ def pytest_unconfigure(config):
                 if os.path.exists(short_filename):
                     lines[os.path.abspath(short_filename)].extend(cdata.lines(filename))
 
-        combined_cdata.add_lines(lines)
 
-        combined_cdata.write_file('.coverage.subprocess')
+        if HAS_COVERAGE >= 5:
+            # Support coverage<5 and >=5; see
+            # https://github.com/astropy/extension-helpers/issues/24
+            # We create an empty coverage data object
+            combined_cdata = CoverageData(suffix='subprocess')
+            combined_cdata.add_lines(lines)
+            combined_cdata.write()
+        else:
+            combined_cdata = CoverageData()
+            combined_cdata.add_lines(lines)
+            combined_cdata.write_file('.coverage.subprocess')

--- a/extension_helpers/conftest.py
+++ b/extension_helpers/conftest.py
@@ -8,7 +8,8 @@ import os
 import glob
 
 try:
-    from coverage import CoverageData, __version__ as coverage_version
+    from coverage import CoverageData
+    from coverage import __version__ as coverage_version
 except ImportError:
     HAS_COVERAGE = False
     CoverageData = None
@@ -50,7 +51,6 @@ def pytest_unconfigure(config):
                 short_filename = filename[pos:]
                 if os.path.exists(short_filename):
                     lines[os.path.abspath(short_filename)].extend(cdata.lines(filename))
-
 
         if HAS_COVERAGE >= 5:
             # Support coverage<5 and >=5; see

--- a/extension_helpers/tests/__init__.py
+++ b/extension_helpers/tests/__init__.py
@@ -4,13 +4,7 @@ import subprocess as sp
 
 import pytest
 
-try:
-    from coverage import CoverageData
-except ImportError:
-    HAS_COVERAGE = False
-else:
-    HAS_COVERAGE = True
-    from ..conftest import SUBPROCESS_COVERAGE
+from ..conftest import HAS_COVERAGE, SUBPROCESS_COVERAGE, CoverageData
 
 PACKAGE_DIR = os.path.dirname(__file__)
 
@@ -65,7 +59,13 @@ def run_setup(setup_script, args):
         stdout, stderr = p.communicate()
 
         cdata = CoverageData()
-        cdata.read_file(os.path.join(path, '.coverage'))
+        if HAS_COVERAGE >= 5:
+            # Support coverage<5 and >=5; see
+            # https://github.com/astropy/extension-helpers/issues/24
+            cdata.read()
+        else:
+            cdata.read_file(os.path.join(path, '.coverage'))
+
         SUBPROCESS_COVERAGE.append(cdata)
 
     else:

--- a/extension_helpers/tests/test_setup_helpers.py
+++ b/extension_helpers/tests/test_setup_helpers.py
@@ -164,6 +164,7 @@ def test_compiler_module(capsys, c_extension_test_package):
 
     with install_temp.as_cwd():
         import apyhtest_eva
+
         # Make sure we imported the apyhtest_eva package from the correct place
         dirname = os.path.abspath(os.path.dirname(apyhtest_eva.__file__))
         assert dirname == str(install_temp.join('apyhtest_eva'))

--- a/setup.cfg
+++ b/setup.cfg
@@ -54,4 +54,5 @@ known_numpy = numpy
 multi_line_output = 0
 balanced_wrapping = True
 include_trailing_comma = false
-length_sort_stdlib = true
+length_sort = False
+length_sort_sections=stdlib

--- a/setup.cfg
+++ b/setup.cfg
@@ -34,7 +34,7 @@ extension_helpers = src/compiler.c
 test =
     pytest
     pytest-cov
-    coverage<5
+    coverage>=4
 docs =
     sphinx
     sphinx-automodapi

--- a/tox.ini
+++ b/tox.ini
@@ -48,4 +48,4 @@ deps =
     isort
 commands =
     pycodestyle extension_helpers --max-line-length=100
-    isort -c -rc extension_helpers
+    isort -c extension_helpers

